build_compilers/lib/src/workers.dart

```diff
@@ -9,76 +9,134 @@ import 'dart:math' show min;
 import 'package:bazel_worker/driver.dart';
 import 'package:build/build.dart';
 import 'package:cli_util/cli_util.dart' as cli_util;
+import 'package:meta/meta.dart';
 import 'package:path/path.dart' as p;
 
 import 'scratch_space.dart';
 
-/// Completes once the analyzer workers have been shut down.
-Future<Null> get analyzerWorkersAreDone =>
-    _analyzerWorkersAreDoneCompleter?.future ?? new Future.value(null);
-Completer<Null> _analyzerWorkersAreDoneCompleter;
+/// On windows some tools have to be executed with the `.bat` extension.
+final String _scriptExtension = Platform.isWindows ? '.bat' : '';
 
-/// Completes once the dartdevc workers have been shut down.
-Future<Null> get dartdevcWorkersAreDone =>
-    _dartdevcWorkersAreDoneCompleter?.future ?? new Future.value(null);
-Completer<Null> _dartdevcWorkersAreDoneCompleter;
+/// Returns the path to the SDK.
+final String sdkDir = cli_util.getSdkPath();
 
-String get _scriptExtension => Platform.isWindows ? '.bat' : '';
+/// Returns the path to a binary in the SDK.
+String _sdkBin(String name) => p.join(sdkDir, 'bin', '$name$_scriptExtension');
 
 final int _defaultMaxWorkers = min((Platform.numberOfProcessors / 2).ceil(), 4);
-
 const _maxWorkersEnvVar = 'BUILD_MAX_WORKERS_PER_TASK';
-
-final int _maxWorkersPerTask = int
-    .parse(Platform.environment[_maxWorkersEnvVar] ?? '$_defaultMaxWorkers',
-        onError: (value) {
-  log.warning('Invalid value for $_maxWorkersEnvVar environment variable, '
-      'expected an int but got `$value`. Falling back to default value '
-      'of $_defaultMaxWorkers.');
-  return _defaultMaxWorkers;
-});
-
-/// Manages a shared set of persistent analyzer workers.
-BazelWorkerDriver get _analyzerDriver {
-  _analyzerWorkersAreDoneCompleter ??= new Completer<Null>();
-  return __analyzerDriver ??= new BazelWorkerDriver(
-      () => Process.start(
-          p.join(sdkDir, 'bin', 'dartanalyzer$_scriptExtension'),
-          ['--build-mode', '--persistent_worker'],
-          workingDirectory: scratchSpace.tempDir.path),
-      maxWorkers: _maxWorkersPerTask);
+final int _maxWorkersPerTask = int.parse(
+  Platform.environment[_maxWorkersEnvVar] ?? '$_defaultMaxWorkers',
+  onError: (value) {
+    log.warning('Invalid value for $_maxWorkersEnvVar environment variable, '
+        'expected an int but got `$value`. Falling back to default value '
+        'of $_defaultMaxWorkers.');
+    return _defaultMaxWorkers;
+  },
+);
+
+/// Encapsulates running an external tool in a sandbox as a compiler.
+///
+/// TODO: Consider making this public in build_compilers(?) in the future.
```

Comment:
I think if we were to do that we'd need to hide details like `WorkResponse`

Reply:
We could change it to a `ProcessResult`, and we could make something that wraps a work response and implements that